buildtest/modules/util.py

```diff
@@ -1,48 +1,156 @@
-import subprocess
+from buildtest.tools.system import BuildTestCommand
 
 
 def get_all_collections():
-    """Return all module collection """
+    """Get all Lmod user collections that is retrieved by running ``module -t savelist``.
+
+     :return: Return all module collections
+     :rtype: list
+     """
+
     collections = "module -t savelist"
-    ret = subprocess.Popen(
-        collections, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE
-    )
-    out = ret.communicate()[1]
-    tmp = out.decode("utf-8").split()
+    cmd = BuildTestCommand()
+    cmd.execute(collections)
+    # The output is captured in STDERR stream.
+    out = cmd.get_error().split()
 
-    return tmp
+    return out
 
 
 class ModuleCollection:
+    """Class declaration of ModuleCollection."""
+
     def __init__(self, collection):
+        """Initializer method of ModuleCollection class.
+
+        :param collection: name of module collection
+        :type collection: str
+        """
+
+        # raise TypeError exception if collection is not a string type since that is required
+        # when working with module collection
+        if not isinstance(collection, str):
+            raise TypeError(f"Type Error: {collection} is not of type string")
+
         self.collection = collection
         self.module_cmd = f"module restore {self.collection}"
 
     def test_collection(self):
-        ret = subprocess.Popen(
-            self.module_cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE
-        )
-        ret.communicate()
-        return ret.returncode
+        """Test the module collection by running ``module restore <collection>``.
```

Comment:
I'm wondering if the ModuleCollection should have similar functions to Module, just of course for the collection? Imagine if we had a list, with a Module and a ModuleCollection:

```python
modules = [Module(), ModuleCollection()]
```

With common methods, we could loop through:

```python
for moduleset in modules:
    moduleset.test()
```
But as it currently is, we'd have to do:

```bash
for moduleset in modules:
    if isinstance(moduleset, ModuleCollection):
        moduleset.testt_collection()
    elif isinstance(moduleset, Module):
        moduleset.test_module()
```
Also, if a module collection truly is a collection of modules, why isn't the Module subclass part of the ModuleCollection? For example, let's say I can create a ModuleCollection and add modules based on a uniqueid:

```
collection = ModuleCollection()
moduleExample = Module(...)
collection["example"] = moduleExample
```
And then the module collection has an internal dict that can be looped over / interacted with in various ways to test.

```bash
def test_collection(self):
    for name, module in self.modules.items():
        print("Testing %s" % name)
        module.test_module()
```
And you could even try subclassing a dict if you wanted the ModuleCollection to act like one. Or have it be a variable like in the example above.

Reply:
The ``ModuleCollection()`` at the moment, was designed to take a collection name so that we can retrieve the command ``module restore <collection>`` this command needs to be in the test script  to load a module collection when buildtest generates a script. 

When i was thinking about module specification in YAML file was to avoid hard coding the module logic in tests as it gets too confusing and complex for framework to handle. So i thought let Lmod do most of the work. 

If ``ModuleCollection()`` is a subclass of ``Module()`` I am not sure how user will interact with this class. 

The ``Module().save()`` and ``Module().describe()`` are two methods that save and show module collection. It is assumed that user can specify which modules to load in the class that get saved. The ``ModuleCollection()`` is assumed to work only with collection that are saved. For instance listing all module collection is just one part of it.

Reply:
I don't mean that ModuleCollection should be a subclass, but rather that ModuleCollection should ultimately load Module objects into it. It's exactly what it sounds like - a collection of actual Modules. This way, when the Module function is developed more, all of that functionality will be available with the ModuleCollection (because it includes Modules somewhere).

Reply:
I'd be open to refactor ModuleCollection class in subsequent PRs. The PR is kinda big and complex. Could we do this in subsequent PRs.

Reply:
Definitely!